### PR TITLE
Refactor greedy algorithms

### PR DESCRIPTION
**Pull request by sdrave**
In this PR, `pymor.algorithms.greedy.greedy` is refactored into a `weak_greedy` function, working with an arbitrary surrogate for the greedy approximation error, and a `rb_greedy` function, which calls `weak_greedy` with an `RBSurrogate`. Similarly, `adaptive_greedy` has been refactored into `adaptive_weak_greedy` and `rb_adaptive_greedy`.

Further changes:

- The old methods `greedy` and `adaptive_greedy` have been deprecated.
- In `rb_greedy`, the `samples` argument has been renamed to `training_set`.

A `strong_greedy` function directly working on `VectorArrays` will follow in the future.

`getting_started.rst` still requires updating.

### CODE REVIEW

File: src/pymor/algorithms/greedy.py [hunk 1]
```diff
@@ -7,22 +7,149 @@
...
+    """Weak greedy basis generation algorithm [BCDDPW11]_.
 
-    This algorithm generates a reduced basis by iteratively adding the
-    worst approximated solution snapshot for a given training set to the
-    reduced basis. The approximation error is computed either by directly
-    comparing the reduced solution to the detailed solution or by using
-    an error estimator (`use_estimator == True`). The reduction and basis
-    extension steps are performed by calling the methods provided by the
-    `reductor` and `extension_algorithm` arguments.
+    This algorithm generates an approximation basis for a given set vectors
```
Comment (ftalbrecht) [inline src/pymor/algorithms/greedy.py]: `set vectors` -> `set of vectors`

File: src/pymor/algorithms/greedy.py [hunk 2]
```diff
@@ -7,22 +7,149 @@
...
+    Dict with the following fields:
+
+        :max_errs:               Sequence of maximum estimated errors during the greedy run.
+        :max_err_mus:            The parameters corresponding to `max_errs`.
+        :extensions:             Number of performed basis extensions.
+        :time:                   Total runtime of the algorithm.
+    """
+    logger = getLogger('pymor.algorithms.greedy.weak_greedy')
+    training_set = list(training_set)
+    logger.info(f'Started greedy search on training set of size {len(training_set)}')
```
Comment (ftalbrecht) [inline src/pymor/algorithms/greedy.py]: Not trying to be picky, but you finish most other log messages with a trailing `.`...

File: src/pymor/algorithms/greedy.py [hunk 3]
```diff
@@ -7,22 +7,149 @@
...
+    tic = time.time()
+    if not training_set:
+        logger.info('There is nothing else to do for an empty training set.')
+        return {'max_errs': [], 'max_err_mus': [], 'extensions': 0,
+                'time': time.time() - tic}
+
+    if pool is None:
+        pool = dummy_pool
+    elif pool is not dummy_pool:
+        logger.info(f'Using pool of {len(pool)} workers for parallel greedy search')
```
Comment (ftalbrecht) [inline src/pymor/algorithms/greedy.py]: s.a.

File: src/pymor/algorithms/greedy.py (hunk 2 above)
Comment (sdrave) [reply src/pymor/algorithms/greedy.py]: Fixed.

File: src/pymor/algorithms/greedy.py (hunk 3 above)
Comment (sdrave) [reply src/pymor/algorithms/greedy.py]: Fixed.

**Comment by sdrave** [comment]
> why rb_adaptive_greedy and not adaptive_rb_greedy? It seems to me the latter would fit the other names a bit better...

I prefer `rb_adaptive_greedy` as it is the rb version (using the rb surrogate) of the `adaptive_greedy` algorithm.

**Comment by sdrave** [comment]
> I prefer rb_adaptive_greedy as it is the rb version (using the rb surrogate) of the adaptive_greedy algorithm.

To be more precise: It should be `rb_adaptive_weak_greedy` but since `rb` implies `weak` I removed that part.

File: src/pymor/algorithms/greedy.py (hunk 1 above)
Comment (sdrave) [reply src/pymor/algorithms/greedy.py]: fixed

**Comment by ftalbrecht** [comment]
> To be more precise: It should be rb_adaptive_weak_greedy but since rb implies weak I removed that part.

Makes sense

**Pull request merged**